perf/light.cc

```diff
@@ -0,0 +1,653 @@
+// pointers to them (since light hash table does not support entries with
+// sizeof >= 8 bytes). So let's use just random one :)
+constexpr static std::size_t TUPLE_SIZE = 1 << 5;
+
+// We measure perf on large datasets: min number of elements is
+// considered to be 10k; max - 1M (which I guess covers the most
+// popular amount of data stored in Tarantool indexes).
+constexpr static std::size_t TUPLE_COUNT_MIN = 10000;
+constexpr static std::size_t TUPLE_COUNT_MAX = 100 * TUPLE_COUNT_MIN;
+constexpr static std::size_t TUPLE_COUNT_MULTIPLIER = 10;
+
+////////////////////////////// Data Definitions ////////////////////////////////////////////////////////////////////////
+
+// 'size' can be skipped in fact (since it is the same for all tuples).
+// Let's keep it just in case (and to make benchmark look a bit more closer
+// to real world). In order to compare two TupleRaw we use memcpy.
+struct TupleRaw {
+	TupleRaw(unsigned char *ptr, std::size_t len) :
+		data(ptr), size(len) { }
+	unsigned char *data;
+	std::size_t size;
+};
+
+using Key_t   = uint8_t;
+using Hash_t  = uint32_t;
+using Value_t = TupleRaw;
+
+// Light hash map implementation does not include any hashing function;
+// it's up to user to choose the proper one. Let's use quite primitive
+// yet good enough: Fowler–Noll–Vo (-1a).
+namespace Hash {
+	static constexpr std::uint32_t offset_basis = 0x811c9dc5;
+	static constexpr std::uint32_t prime = 0x01000193;
+
+	Hash_t
+	FNV_hash(const unsigned char *ptr, std::size_t len)
+	{
+		std::uint32_t hash = offset_basis;
+		for (std::size_t i = 0; i < len; ++i) {
+			hash ^= (unsigned char) ptr[i];
+			hash *= prime;
+		}
+		return hash;
+	}
+};
+
+// Wrapper to move hash calculation to setup stage (in order to make
+// performance benchmark results cleaner).
+struct Tuple {
+	Tuple(unsigned char *ptr, std::size_t len, uint32_t key) :
+		raw(ptr, len), key(key), hash(Hash::FNV_hash(ptr, len)) { }
+	Value_t raw;
+	Key_t key;
+	Hash_t hash;
+};
+
+namespace Hash {
+	// Since std::unordered_set does not provide an ability to specify
+	// hash value beforehand, let's use struct Tuple as values with
+	// pre-calculated hashes (yes, unordered_set stores hash values on its
+	// own but we have nothing left to do here to make it closer to
+	// Light bench).
+	struct TupleHash {
+		std::size_t operator()(const Tuple &t) const
+		{
+			return t.hash;
+		}
+	};
+}
+
+// Simple generator of uniformly distributed random bytes.
+struct RandomBytesGenerator {
+	RandomBytesGenerator() :
+		gen(std::random_device()()), distr(0, UCHAR_MAX) { }
+	unsigned char get() { return distr(gen); }
+	std::mt19937 gen;
+	std::uniform_int_distribution<unsigned short> distr;
+};
+
+// Large chunk of continuous memory initialized with random bytes.
+struct TupleHolder {
+	TupleHolder(std::size_t tuple_count) : mem(nullptr), tuples()
+	{
+		std::size_t total_size = tuple_count * TUPLE_SIZE;
+		mem = static_cast<unsigned char *>(malloc(total_size));
```

Comment:
Please don't use `malloc` in C++ code.

Reply:
Ok)

Reply:
Why?. note that `new char[N]` is a bit more expensive.

Reply:
Discussed f2f, seems that it's not vital for this benchmark (as it doesn't affect measurements)

Reply:
@alyapunov `operator new` and `operator delete` are simply wrappers over `malloc` and `free` (see for instance https://github.com/llvm/llvm-project/blob/main/libcxx/src/new.cpp).